bin/Macronify/Main.cpp

```diff
@@ -0,0 +1,734 @@
+                    builder->getBoolAttr(function_like),
+                    return_type,
+                    std::move(region)
+                );
+            } else {
+                return StmtVisitor::template make<macroni::MacroParameter>(
+                    loc,
+                    builder->getStringAttr(llvm::Twine(macro_name)),
+                    return_type,
+                    std::move(region)
+                );
+            }
+        }
+
+        Op *VisitNonMacro(const clang::Stmt *stmt) {
+            auto op = StmtVisitor::Visit(stmt);
+            if (vast::hl::CallOp call_op =
+                mlir::dyn_cast<vast::hl::CallOp>(op)) {
+                auto name = call_op.getCalleeAttr().getValue();
+                if ("rcu_read_lock" == name) {
+                    call_op.getOperation()->setAttr(
+                        "lock_level",
+                        builder->getI64IntegerAttr(lock_level));
+                    lock_level++;
+                } else if ("rcu_read_unlock" == name) {
+                    lock_level--;
+                    call_op.getOperation()->setAttr(
+                        "lock_level",
+                        builder->getI64IntegerAttr(lock_level));
+                }
+            }
+            return op;
+        }
+    };
+
+    template< typename Derived >
+    using VisitorConfig = vast::cg::CodeGenFallBackVisitorMixin< Derived,
+        CodeGenVisitorMixin,
+        vast::cg::DefaultFallBackVisitorMixin
+    >;
+
+    using Visitor =
+        vast::cg::CodeGenVisitor<VisitorConfig, MacroniMetaGenerator>;
+
+} // namespace macroni
+
+static llvm::cl::list< std::string > compiler_args(
+    "ccopts", llvm::cl::ZeroOrMore, llvm::cl::desc("Specify compiler options")
+);
+
+int main(int argc, char **argv) {
+
+    pasta::InitPasta initializer;
+    pasta::FileManager fm(pasta::FileSystem::CreateNative());
+    auto maybe_compiler =
+        pasta::Compiler::CreateHostCompiler(fm, pasta::TargetLanguage::kCXX);
+    if (!maybe_compiler.Succeeded()) {
+        std::cerr << maybe_compiler.TakeError() << std::endl;
+        return EXIT_FAILURE;
+    }
+
+    auto maybe_cwd = (pasta::FileSystem::From(maybe_compiler.Value())
+                      ->CurrentWorkingDirectory());
+    if (!maybe_cwd.Succeeded()) {
+        std::cerr << maybe_compiler.TakeError() << std::endl;
+        return EXIT_FAILURE;
+    }
+
+    const pasta::ArgumentVector args(argc - 1, &argv[1]);
+    auto maybe_command = pasta::CompileCommand::CreateFromArguments(
+        args, maybe_cwd.TakeValue());
+    if (!maybe_command.Succeeded()) {
+        std::cerr << maybe_command.TakeError() << std::endl;
+        return EXIT_FAILURE;
+    }
+
+    const auto command = maybe_command.TakeValue();
+    auto maybe_jobs = maybe_compiler->CreateJobsForCommand(command);
+    if (!maybe_jobs.Succeeded()) {
+        std::cerr << maybe_jobs.TakeError() << std::endl;
+        return EXIT_FAILURE;
+    }
+
```

Comment:
Up to this point you are only doing sanity checking. I would give it a separate function, that simply returns object/tuple of all relevant things (maybe as an optional so you can communicate failure).